File: 15-3Sum/15-3Sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        set<vector<int>> triplets;
        for (int i = 0; i < nums.size(); ++i) {
            int l = i + 1, r = nums.size() - 1;
            while (l < nums.size() && l < r) {
                long int total = nums.at(i) + nums.at(l) + nums.at(r);
                if (total == 0) {
                    triplets.insert({nums.at(i), nums.at(l), nums.at(r)});
                    l++;
                    r--;
                }
                else if (total < 0) {
                    l++;
                }
                else {
                    r--;
                }
            }
        }
        return vector<vector<int>>(triplets.begin(), triplets.end());
    }
};

File: 15-3Sum/15-3Sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        triplets = set()
        n = len(nums)
        for i in range(n):
            l, r = i + 1, n - 1
            while (l < n and l < r):
                target = nums[i] + nums[l] + nums[r]
                if (nums[i] + nums[l] + nums[r]) == 0:
                    triplets.add((nums[i], nums[l], nums[r]))
                    l += 1
                    r -= 1
                elif target < 0:
                    l += 1
                else:
                    r -= 1
        return [list(x) for x in triplets]